pandas/io/_plugin_loader.py

```diff
@@ -0,0 +1,176 @@
+    the connector, since the return of third-party connectors is not necessarily
+    a pandas DataFrame but any object supporting the dataframe interchange
+    protocol. We make sure here that `read_<whatever>` returns a pandas DataFrame.
+    """
+
+    # TODO: Create this function dynamically so the resulting signature contains
+    # the original parameters and not `*args` and `**kwargs`
+    @functools.wraps(io_plugin.reader)
+    def reader_wrapper(*args, **kwargs):
+        result = io_plugin.reader(*args, **kwargs)
+
+        if isinstance(result, list):
+            result = [pd.api.interchange.from_dataframe(df) for df in result]
+        elif isinstance(result, dict):
+            result = {
+                k: pd.api.interchange.from_dataframe(df) for k, df in result.items()
+            }
+        else:
+            result = pd.api.interchange.from_dataframe(result)
+
+        return result
+
+    # TODO `function.wraps` changes the name of the wrapped function to the
+    # original `pandas_reader`, change it to the function exposed in pandas.
+    return reader_wrapper
+
+
+def _create_series_writer_function(format_name):
+    """
+    When calling `Series.to_<whatever>` we call the dataframe writer, so
+    we need to convert the Series to a one column dataframe.
+    """
+
+    def series_writer_wrapper(self, *args, **kwargs):
+        dataframe_writer = getattr(self.to_frame(), f"to_{format_name}")
+        dataframe_writer(*args, **kwargs)
+
+    return series_writer_wrapper
+
+
+def _warn_conflict(func_name, format_name, loaded_plugins, module):
+    package_to_load = importlib_metadata.packages_distributions()[module.__name__]
+    if format_name in loaded_plugins:
+        # conflict with a third-party connector
+        loaded_module = loaded_plugins[format_name]
+        loaded_package = importlib_metadata.packages_distributions()[
+            loaded_module.__name__
+        ]
+        msg = (
+            f"Unable to create `{func_name}`. "
+            f"A conflict exists, because the packages `{loaded_package}` and "
+            f"`{package_to_load}` both provide the connector for the '{format_name}' format. "
+            "Please uninstall one of the packages and leave in the current "
+            "environment only the one you want to use for the '{format_name}' format."
+        )
+    else:
+        # conflict with a pandas connector
+        msg = (
+            f"The package `{package_to_load}` registers `{func_name}`, which is "
+            "already provided by pandas. The plugin will be ignored."
+        )
+
+    warnings.warn(msg, UserWarning, stacklevel=1)
+
+
+def load_io_plugins():
+    """
+    Looks for entrypoints in the `dataframe.io` group and creates the
+    corresponding pandas I/O methods.
+    """
+    loaded_plugins = {}
+
+    for dataframe_io_entry_point in entry_points().get("dataframe.io", []):
+        format_name = dataframe_io_entry_point.name
```

Comment:
Where does this name get defined? Assuming from the name of the library itself? If so maybe worth making this a property of the class so that there is some flexibility for package authors

Reply:
This is the name of the entrypoint. Package authors define it explicitly as the name pandas will use in `read_<name>`... It's not use for anything else. The only constrain is that the name Dask, Vaex, Polars... Will receive if they ever use this connector API will be the same. Personally I think that's good, but not sure if for any case the same connector would want to use different names in different libraries.

Reply:
Is there any validity to one package providing multiple read/write implementations? An example might be excel where one package offers read_xls alongside read_xlsx, etc...

Reply:
There shouldn't be any limitation about that